toolchain/check/handle_require.cpp

```diff
@@ -86,16 +87,21 @@ auto HandleParseNode(Context& context, Parse::RequireDefaultSelfImplsId node_id)
 auto HandleParseNode(Context& context, Parse::RequireTypeImplsId node_id)
     -> bool {
   auto [self_node_id, self_inst_id] = context.node_stack().PopExprWithNodeId();
+  auto self_type = ExprAsType(context, self_node_id, self_inst_id);
 
-  auto introducer = context.decl_introducer_state_stack().innermost();
+  const auto& introducer = context.decl_introducer_state_stack().innermost();
   if (introducer.modifier_set.HasAnyOf(KeywordModifierSet::Extend)) {
-    CARBON_DIAGNOSTIC(RequireImplsExtendWithExplicitSelf, Error,
-                      "`extend require impls` with explicit type");
-    context.emitter().Emit(self_node_id, RequireImplsExtendWithExplicitSelf);
-    self_inst_id = SemIR::ErrorInst::InstId;
+    if (self_type.type_id != SemIR::ErrorInst::TypeId) {
+      CARBON_DIAGNOSTIC(RequireImplsExtendWithExplicitSelf, Error,
+                        "`extend require impls` with explicit type");
+      // TODO: If the explicit self-type matches a lookup of NameId::SelfType,
+      // add a note to the diagnostic: "remove the explicit `Self` type here",
+      // and continue without an ErrorInst. See ExtendImplSelfAsDefault.
+      context.emitter().Emit(self_node_id, RequireImplsExtendWithExplicitSelf);
```

Comment:
It seems a bit inconsistent to be diagnosing this here but diagnosing the class case when handling the `ImplTypeAs` node.

Reply:
I am not following this comment, as I was trying to make these two be consistent here. Maybe the names are being confusing, or maybe I am very confused so here's what I think.

Both of these are now diagnosing in the place where they are parsing an explicit self-type, but see it's `extend` so they shouldn't be in that handler:
- RequireTypeImplsId is the node for an explicit self-type `require <here> impls I`
- ImplTypeAsId is the node for an explicit self-type `impl <here> as I`

Both of these functions are where we determine the self, possibly overwrite it to an error, and push it to the node stack.